File: exercism/ruby/raindrops/raindrops.rb
class Raindrops
  MAP = {3 => 'Pling', 5 => 'Plang', 7 => 'Plong'}

  def self.convert(n)
    mapped = factorize(n).map { |n| MAP[n] }.compact
    mapped.empty? ? n.to_s : mapped.join("")
  end

  def self.factorize(n)
    primes, i, limit = [], 2, n

    while i <= limit
      if n % i == 0
        primes << i
        n = n / i
      else
        i = i + 1
      end
    end

    primes.uniq
  end
end

File: exercism/scala/raindrops/src/main/scala/raindrops.scala
case class Raindrops() {
  def convert(n: Int): String = {
    var out = ""

    if ((n % 3) == 0) out += "Pling"
    if ((n % 5) == 0) out += "Plang"
    if ((n % 7) == 0) out += "Plong"

    if (out == "") n.toString else out
  }
}
